Finish SQLProduct migration

emord opened the pull request:
Just opening for the sake of public work. builds off of https://github.com/dimagi/commcare-hq/pull/25157/. I think it should be possible to complete this next week. Its been in progress for nearly 5 years (> 20k PRs have passed since it was first introduced) https://github.com/dimagi/commcare-hq/pull/4043

- [x] SQLProuduct support for from_excel
- [x] SQLProduct support archive/unarchive
- [x] save directly to SQLProduct (bypass couch)
- [x] move make_product in tests to SQLProduct only

emord committed:
Remove Product use from commtrack tests

emord committed:
Remove unused Product methods

emord committed:
Standardize on one make_product

emord committed:
Remove unused class properties

emord committed:
Change report from Product to SQLProduct

emord committed:
Convert calls from Product -> SQLProduct

emord committed:
Convert product download to SQLProduct

millerdev reviewed (commented):
🏅 Looking great!

emord committed:
Merge branch 'master' into je/sql-products-v2

emord committed:
Use exists

emord committed:
Use active_objects to filter archived products

emord committed:
Assign attributes during object creation

emord committed:
Support archive/unarchive SQLProducts

emord committed:
Update product tests that used archive

millerdev reviewed (commented):
⬅️ (review progress marker)

emord commented on corehq/apps/products/forms.py:
```diff
@@ -65,8 +65,13 @@ def __init__(self, product, *args, **kwargs):
+            SQLProduct.objects
+            .filter(domain=self.product.domain, name=name)
+            .exclude(product_id=self.product._id)
+        ).count()
```
Nice. I forgot about `exists` https://github.com/dimagi/commcare-hq/pull/25194/commits/7758abb9f12ff646f17c27e213daa4b009194de8

emord commented on corehq/apps/commtrack/tests/util.py:
```diff
@@ -103,12 +103,13 @@ def bootstrap_location_types(domain):
     p.name = name
     p.code = code.lower()
-    p.program_id = program_id
+    if program_id:
```
there are some tests with it `''`and that's fine. i ended up changing this to be nicer anyways https://github.com/dimagi/commcare-hq/pull/25194/commits/64ec64db464bdc8a886f9c46eacac6607e270a7e

emord commented on custom/ewsghana/reports/maps.py:
```diff
@@ -34,6 +34,13 @@ def active_location(self):
+    def active_product(self):
+        product_id = self.config.get('product_id')
+        if product_id:
+            return SQLProduct.objects.get(domain=self.domain, product_id=product_id)
```
looks like i accidentally pushed this in a separate commit to use active_objects https://github.com/dimagi/commcare-hq/pull/25194/commits/b6215f78cce866f16ae45019ef7f71172409dfbd#diff-2c50b07a49c843fb5c5fd3eeab0e5b5aR41

emord committed:
Fix tests

emord committed:
Standardize tests to use make_product

emord committed:
API uses SQLProduct

emord committed:
Remove unneeded management command

emord committed:
Use make_product

esoergel reviewed (commented):
This is fantastic!  In defense of whoever created `SQLProduct`, I think it was originally intended as a read-only clone for reporting purposes, similar to how we use ES.  We're much better off strictly on SQL, IMO though, so I'm very happy to see this progress.

emord committed:
Merge branch 'master' into je/sql-products-v2

emord committed:
Fix six conflict

emord committed:
Set is_archived on current instance

emord committed:
Use _id instead of product_id for Group

emord committed:
Force query to evaluate in test

emord committed:
Fix product name

emord committed:
Support SQLProduct delte

emord committed:
Remove unused product caching code

emord committed:
Import into SQLProduct

emord committed:
Update SQLProduct directly on program change

emord committed:
Save directly to SQLProduct

emord commented on custom/ewsghana/tests/ews_test_reminders.py:
```diff
@@ -21,12 +22,8 @@ def setUpClass(cls):
-        self.commodity2 = Product(domain=TEST_DOMAIN, name='Drug B', code_='cd', unit='cycle')
-        self.commodity2.save()
-
+        self.commodity = make_product(TEST_DOMAIN, 'Drug A', '_ab', unit='cycle')
```
No. it also caused test errors https://github.com/dimagi/commcare-hq/pull/25194/commits/e33e2bbb38d70edca0abe4be42b8557ee875e0c6

emord commented:
> In defense of whoever created SQLProduct, I think it was originally intended as a read-only clone for reporting purpose

Yeah that's been pretty clear on these last few steps to get it over the hump.

Anyways, I believe that this is ~done functionality wise. the only remaining reference to `Product` is in the `clone_domain` mgmt command (I thought it wouldn't even work anymore, but there are tests around it so that makes think it does work). I don't totally trust our test suite to test the import and editing/creation pages, so I'll at least do some manual tests on those.

This is a big enough change that I'd rather do it after I get back from PTO on the 23rd unless someone else particularly wants to see it through

emord committed:
Add DeprecationWarnings

emord committed:
Merge branch 'master' into je/sql-products-v2

emord commented:
had some errors from ewsghana tests, but those are now removed, so I merged master. Will look to have this done with tests passing by end of week (or at least pull out the easy test changes into a separate PR to reduce the size of this one)

emord committed:
Remvoe Product from couch dump

emord committed:
Remove couch products from clone_domain

emord committed:
Remove Product model

emord committed:
Remove more couch code from the models file

emord committed:
Merge branch 'master' into je/sql-products-v2

emord committed:
Add Product to deleteable doc type

emord commented:
definitely regret not pulling these out into separate PRs earlier (before ewsghana was removed). Will plan to release next week after thanksgiving since i'm offline most of hte week

emord committed:
Merge branch 'master' into je/sql-products-v2

emord committed:
Save SQLProduct in fixture generator test

Saving the SQLProduct forces the floats to be converted to Decimal and
is a better test of what would happen in a real fixture

snopoke commented on corehq/apps/products/models.py:
```diff
@@ -219,123 +154,49 @@ def from_excel(cls, row, custom_data_validator):
         if id:
             try:
-                p = cls.get(id)
+                product = cls.objects.get(location_id=id)
```
shouldn't this be `product_id=id`?

snopoke reviewed (commented):
One question but otherwise looks good

emord committed:
Change product.doc to dict

emord committed:
Fix product from_excel

emord commented:
I tested this out on staging for a bit. Most things appeared to work at first blush, but the errors I ran into made me super cautious about this (particularly the late one where a form used the [hidden `_doc` to get a dict](https://github.com/dimagi/commcare-hq/pull/25194/commits/9060b0e8ba8e055e2b2a1698af8e14abc04cb1e1) )

The things I found that are definitely untested are:
* the forms (both updated and created)
* exports

Those should probably get tests added before a final migration to ensure that the various private interfaces used here are not going to cause issues
